app/services/external_upload_manager.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+class ExternalUploadManager
+  DOWNLOAD_LIMIT = 100.megabytes
+
+  class ChecksumMismatchError < StandardError; end
+  class DownloadFailedError < StandardError; end
+  class CannotPromoteError < StandardError; end
+
+  attr_reader :external_upload_stub
+
+  def initialize(external_upload_stub)
+    @external_upload_stub = external_upload_stub
+  end
+
+  def can_promote?
+    external_upload_stub.status == ExternalUploadStub.statuses[:created]
+  end
+
+  def promote_to_upload!
+    raise CannotPromoteError if !can_promote?
+
+    external_upload_stub.update!(status: ExternalUploadStub.statuses[:uploaded])
+    external_stub_object = Discourse.store.object_from_path(external_upload_stub.key)
+    external_etag = external_stub_object.etag
+    external_size = external_stub_object.size
+    external_sha1 = external_stub_object.metadata["sha1-checksum"]
+
+    # This could be legitimately nil, if it's too big to download on the
+    # server, or it could have failed. To this end we set a should_download
+    # variable as well to check.
+    tempfile = nil
+    should_download = external_size < DOWNLOAD_LIMIT
+    if should_download
+      tempfile = download(external_upload_stub.key, external_upload_stub.upload_type)
+
+      raise DownloadFailedError if tempfile.blank?
+
+      actual_sha1 = Upload.generate_digest(tempfile)
+      if external_sha1 && external_sha1 != actual_sha1
+        raise ChecksumMismatchError
+      end
+    end
```

Comment:
I have some concerns about this. I thought the point of direct to S3 uploads is so that the file will never have to pass through our servers. The limit is quite huge so multiple requests downloading files at the limit can easily exhaust the resources on our server.

Reply:
Ahh icic. The limit is similar to the max limit we have on our site settings. I guess this is fine then.

Reply:
O wait, actually I still don't get it. If we eventually download files below a certain size back on to our server, why don't we just upload it directly to the server in the first place.

Reply:
We resolved this after discussions in chat :)